packages/@lwc/engine-core/src/framework/mutation-logger.ts

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) 2024, Salesforce, Inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+//
+// Do additional mutation tracking for DevTools performance profiling, in dev mode only.
+//
+import {
+    ArrayPush,
+    ArraySplice,
+    isUndefined,
+    toString,
+    isObject,
+    isNull,
+    isArray,
+    keys,
+} from '@lwc/shared';
+import { ReactiveObserver } from '../libs/mutation-tracker';
+import { VM } from './vm';
+import { assertNotProd } from './utils';
+
+export interface MutationLog {
+    vm: VM;
+    prop: string;
+}
+
+const reactiveObserversToVMs = new WeakMap<ReactiveObserver, VM>();
+const targetsToPropertyKeys = new WeakMap<object, PropertyKey>();
+let mutationLogs: MutationLog[] = [];
+
+/**
+ * Flush all the logs we've written so far and return the current logs.
+ */
+export function getAndFlushMutationLogs() {
+    assertNotProd();
+    const result = mutationLogs;
+    mutationLogs = [];
+    return result;
+}
+
+/**
+ * Log a new mutation for this reactive observer.
+ * @param reactiveObserver - relevant ReactiveObserver
+ * @param target - target object that is being observed
+ * @param key - key (property) that was mutated
+ */
+export function logMutation(reactiveObserver: ReactiveObserver, target: object, key: PropertyKey) {
+    assertNotProd();
+    const parentKey = targetsToPropertyKeys.get(target);
+    const vm = reactiveObserversToVMs.get(reactiveObserver);
+
+    /* istanbul ignore if */
+    if (isUndefined(vm)) {
+        // VM should only be undefined in vitest tests, where a reactive observer is not always associated with a VM
+        // because the unit tests just create Reactive Observers on-the-fly.
+        if (process.env.NODE_ENV === 'test-karma-lwc') {
```

Comment:
Nope, because I don't want this throwing Errors in our downstream consumer's Jest/Vitest code. :) Whereas with `test-karma-lwc` we know it's only ever going to throw in our own tests.

I can add a comment!